src/seeds.gratitude.cpp

```diff
@@ -0,0 +1,197 @@
+#include <seeds.gratitude.hpp>
+
+
+ACTION gratitude::reset () {
+  require_auth(get_self());
+
+  auto bitr = balances.begin();
+  while (bitr != balances.end()) {
+    bitr = balances.erase(bitr);
+  }
+
+  auto sitr = sizes.begin();
+  while (sitr != sizes.end()) {
+    sitr = sizes.erase(sitr);
+  }
+
+  auto stitr = stats.begin();
+  while (stitr != stats.end()) {
+    stitr = stats.erase(stitr);
+  }
+
+  // setup first round
+  stats.emplace(_self, [&](auto& item) {
+    item.round_id = 1;
+    item.num_transfers = 0;
+    item.volume = asset(0, gratitude_symbol);
+  });
+
+}
+
+ACTION gratitude::give (name from, name to, asset quantity, string memo) {
+  require_auth(from);
+
+  check( from != to, "gratitude: cannot give to self" );
+  check( is_account( to ), "gratitude: to account does not exist");
+  check_asset(quantity);
+  check( quantity.amount > 0, "gratitude: must give positive quantity" );
+
+  // Should create balances if not there yet
+  init_balances(to);
+  init_balances(from);
+
+  sub_gratitude(from, quantity);
+  add_gratitude(to, quantity);
+
+  update_stats(from, to, quantity);
+}
+
+ACTION gratitude::newround() {
+  require_auth(get_self());
+  auto generated_gratz = config_get(gratzgen);
+
+  auto contract_balance = eosio::token::get_balance(contracts::token, get_self(), seeds_symbol.code());
+  uint64_t tot_accounts = get_size("balances.sz"_n);
+  uint64_t volume = get_current_volume();
+
+
+  auto balances_by_received = balances.get_index<"byreceived"_n>();
+
+  auto bitr = balances_by_received.begin();
+  while (bitr != balances_by_received.end()) {
+    uint64_t my_received = bitr->received.amount;
+    balances_by_received.modify(bitr, _self, [&](auto& item) {
+        item.received = asset(0, gratitude_symbol);
+        item.remaining = asset(generated_gratz, gratitude_symbol);
+    });
+    float split_factor = my_received / (float)volume;
+    uint64_t payout = contract_balance.amount * split_factor;
+    // Pay out SEEDS in store
+    if (payout > 0) _transfer(bitr->account, asset(payout, seeds_symbol), "gratitude bonus");
+    bitr++;
+  }
+}
+
+/// ----------================ PRIVATE ================----------
+
+void gratitude::check_user (name account) {
+  auto uitr = users.find(account.value);
+  check(uitr != users.end(), "gratitude: user not found");
+}
+
+uint64_t gratitude::get_current_volume() {
+  auto stitr = stats.rbegin();
+  // Should always work because reset always creates first round
+  return stitr->volume.amount;
+}
+
+void gratitude::update_stats(name from, name to, asset quantity) {
+  auto stitr = stats.rbegin();
+  auto round_id = stitr->round_id;
+
+  auto stitr2 = stats.find(round_id);
+  auto oldvolume = stitr2->volume.amount;
```

Comment:
stitr2 is now == stitr - why not just use sitr?

I don't understand this code...

Reply:
This is to find the `round_id` of the last round (`stats.rbegin()`), so that's why I need 2 iterators.